src/transformers/benchmark/benchmark_utils_generic.py

```diff
@@ -0,0 +1,111 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Benchmark
+"""
+import json
+import timeit
+
+
+class BenchMark:
+    def __init__(self, *arg, **kwargs):
+        self._buffer = {
+            "init_kwargs": {},
+            "inputs_kwargs": {},
+            "target_kwargs": {},
+            "measure_kwargs": {},
+            "report_kwargs": {},
+        }
+
+    def _reset_buffer(self):
+        self._buffer["inputs_kwargs"] = {}
+        self._buffer["target_kwargs"] = {}
+        self._buffer["measure_kwargs"] = {}
+        self._buffer["report_kwargs"] = {}
+
+    def measure(self, func, **measure_kwargs):
+        raise NotImplementedError
+
+    def target(self, **target_kwargs):
+        raise NotImplementedError
+
+    def inputs(self, **inputs_kwargs):
+        return {}
+
+    def _convert_to_json(self, report):
+        if isinstance(report, list):
+            return [self._convert_to_json(x) for x in report]
+        if isinstance(report, dict):
+            return {k: self._convert_to_json(v) for k, v in report.items()}
+        if isinstance(report, type):
+            return report.__name__
+        return report
+
+    def report(self, result, only_result=False, output_path=None):
+        report = {"result": result}
+        if not only_result:
+            report["configuration"] = self._buffer
+
+        report = self._convert_to_json(report)
+        if output_path is not None:
+            with open(output_path, "w", encoding="UTF-8") as fp:
+                json.dump(report, fp, ensure_ascii=False, indent=4)
+
+        return report
+
+    def run(self, measure_kwargs=None, target_kwargs=None, inputs_kwargs=None, report_kwargs=None):
+        self._reset_buffer()
+
+        if measure_kwargs is None:
+            measure_kwargs = {}
+        if target_kwargs is None:
+            target_kwargs = {}
+        if inputs_kwargs is None:
+            inputs_kwargs = {}
+        if report_kwargs is None:
+            report_kwargs = {}
+        if measure_kwargs is None:
+            measure_kwargs = {}
+
+        target = self.target(**target_kwargs)
+
+        all_inputs_kwargs = [inputs_kwargs] if isinstance(inputs_kwargs, dict) else inputs_kwargs
+        results = []
+        for _inputs_kwargs in all_inputs_kwargs:
+            inputs = self.inputs(**_inputs_kwargs)
+            result = self.measure(target, **measure_kwargs)(**inputs)
+            results.append(result)
+
+        if isinstance(inputs_kwargs, dict):
+            results = results[0]
+
+        return self.report(results, **report_kwargs)
+
+
+class SpeedBenchMark(BenchMark):
+    def measure(self, func, number=3, repeat=1):
+        self._buffer["measure_kwargs"]["number"] = number
+        self._buffer["measure_kwargs"]["repeat"] = repeat
```

Comment:
This is a bit funny - someone can init the class with "measure_kwargs" and then they'll always be over written with these defaults silently. 

In fact, even if I called

```benchmark.measure(func, number=5, repeat=2)```

I'm not sure that I would expect it to over write the class defaults s.t. if I then called

```benchmark.measure(func)```

`number=5` and `repeat=2` would be used. 

I'd instead: 
* Enable easily updating the kwargs - so something like `set_kwargs` or `update_kwargs` alongside the `reset_kwargs` argument
* Use the class' values by default, which can be overridden (but not set) by calling the method.

I realise that the intention is that `measure` will be called with `measure_kwargs` but as it's a public method, it's not guaranteed this will happen

Reply:
@amyeroberts All these methods are not to be used directly. Only the `run` (defined in the parent class `Benchmark`) method should be called.

Reply:
In this case, they should all be private

Reply:
ok, I will change them